Snowpark integration tutorial (#2909)
astronomer/docs: learn/airflow-snowpark.md

---
title: "Orchestrate Machine Learning in Snowpark with Apache Airflow"

> sidebar_label: "Snowpark"

Reviewer: ```suggestion
sidebar_label: "Snowpark Tutorial"
```
As we chatted about this morning, let's put this in the Snowflake menu.

description: "Learn how to integrate Snowpark and Airflow."
id: airflow-snowpark
sidebar_custom_props: { icon: 'img/integrations/snowflake.png' }
---

import CodeBlock from '@theme/CodeBlock';
import airflow_with_snowpark_tutorial from '!!raw-loader!../code-samples/dags/airflow-snowpark/airflow_with_snowpark_tutorial.py';

[Snowpark](https://www.snowflake.com/en/data-cloud/snowpark/) is a framework containing runtimes and libraries to run non-SQL code in [Snowflake](https://www.snowflake.com/en/). Snowpark comes with a [comprehensive machine learning library](https://docs.snowflake.com/en/developer-guide/snowpark-ml/index) optimized for Snowflake. 

> In this tutorial you'll learn how to: 

Reviewer: ```suggestion
In this tutorial, you'll learn how to: 
```

> - Create a [custom XCom backend](custom-xcom-backends-tutorial.md) in Snowflake.

Reviewer: Somewhere in these bullets we should have a "...to implement a model predicting XYZ" - however you want to say it, but should generally describe the use case so people understand what the plot is showing :)

- Create and use a model registry in Snowflake.
- Use Airflow decorators to run code in Snowpark, both in a pre-built and custom virtual environment.

:::caution

The provider used in this tutorial is currently in beta and its contents as well as the decorators used are subject to change. After the official release, this tutorial will be updated.

:::

![A plot showing a confusion matrix and ROC curve with a high AUC for hot_chocolate, a medium AUC for snow_mocha and tea and low predictive power for wine and coffee.](/img/tutorials/airflow-snowpark_plot_metrics.png)

## Why use Airflow with Snowpark?

> Snowpark allows you to use Python to perform transformations as well as machine learning operations on data stored in Snowflake.

Reviewer: ```suggestion
Snowpark allows you to use Python to perform transformations and machine learning operations on data stored in Snowflake.
```

Integrating Snowpark with Airflow offers the benefits of:

- Running machine learning models directly in Snowflake, without having to move data out of your Snowflake database.
- Express data transformations in Snowflake in Python instead of SQL.
- Store and version your machine learning models in a model registry inside Snowflake.
- Use Snowpark's compute resources instead of the resources of your Airflow cluster for machine learning.

> Additionally this tutorial shows how to use Snowflake as a custom XCom backend, this is especially useful for organizations with strict compliance requirements who want to keep all their data in Snowflake but still leverage [Airflow XCom](airflow-passing-data-between-tasks.md) to pass data between tasks.

Reviewer: ```suggestion
Additionally, this tutorial shows how to use Snowflake as a custom XCom backend. This is especially useful for organizations with strict compliance requirements who want to keep all their data in Snowflake but still leverage [Airflow XCom](airflow-passing-data-between-tasks.md) to pass data between tasks.
```

## Time to complete

This tutorial takes approximately 45 minutes to complete.

## Assumed knowledge

To get the most out of this tutorial, make sure you have an understanding of:

- The basics of Snowflake and Snowpark. See [Introduction to Snowflake](https://docs.snowflake.com/en/user-guide-intro.html) and the [Snowpark API documentation](https://docs.snowflake.com/en/developer-guide/snowpark/index).
- Airflow decorators. [Introduction to the TaskFlow API and Airflow decorators](airflow-decorators.md).
- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).
- Setup/ teardown tasks in Airflow. See [Use setup and teardown tasks in Airflow](airflow-setup-teardown.md).

## Prerequisites

- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).
- A Snowflake account. A [30-day free trial](https://signup.snowflake.com/) is available. You will need to have at least one database and one schema created to store the data and models used in this tutorial.

> If you want to run this tutorial using the Snowflake custom XCom backend you will need to either:

Reviewer: ```suggestion
If you want to run this tutorial using the Snowflake custom XCom backend, you will need to either:
```

Reviewer: Is using the XCom backend optional? If so, I would lean towards pulling this into a bullet in the list with something like [Optional] To use the Snowflake custom XCom backend, you will need either: and then move the two bullets to subullets.

- Use a Snowflake account with `ACCOUNTADMIN` privileges. In this case the Snowflake custom XCom backend is created and cleaned up by the [setup/ teardown](airflow-setup-teardown.md) tasks (`create_snowflake_objects` and `cleanup_xcom_table`) at the beginning and end of the tutorial DAG. The free trial account does have the required privileges.
- Ask your Snowflake administrator to create the following for you:
    - A database called `SNOWPARK_XCOM_DB`.
    - A schema called `SNOWPARK_XCOM_SCHEMA`.
    - A table called `XCOM_TABLE`.
    - A stage called `XCOM_STAGE`.

> Similarly, the tutorial DAG includes a toggle to use a [Snowpark-optimized warehouse](https://docs.snowflake.com/en/user-guide/warehouses-snowpark-optimized) for model training. While the tutorial DAG uses a small dataset where model training can be accomplished using the standard Snowflake warehouse, Astronomer recommends to use a Snowpark warehouse for model training in production. If you want to use a Snowpark warehouse for model training you will need to either:

Reviewer: I'm leaning towards leaving this out of the prereqs section and just putting an info box somewhere below that while the DAG defaults to using the smaller warehouse (if it doesn't already, we should make it), we recommend in prod using the snowpark optimized one, and note that you need account admin privileges. 

Just to simplify this section a bit. Right now it seems a little overwhelming, so if we can make the defaults simpler and present these things as optional later on I think that would help.

- Ask your Snowflake administrator to create a Snowpark-optimized warehouse (`SNOWPARK_WH`) for you.

:::info

The example code from this tutorial is also available on [GitHub](https://github.com/astronomer/airflow-snowpark-tutorial). 

:::

## Step 1: Configure your Astro project

1. Create a new Astro project:

    ```sh
    $ mkdir astro-snowpark-tutorial && cd astro-snowpark-tutorial
    $ astro dev init
    ```

2. Download the `whl` file for the Astro Snowflake provider beta version from the [Astronomer Github repository](https://github.com/astronomer/learn-tutorials-data/blob/main/wheel_files/astro_provider_snowflake-0.0.0-py3-none-any.whl) and save it in your Astro project's `include` directory.

3. Create a new file in your Astro project's root directory called `requirements-snowpark.txt`. This file contains all Python packages that you will install in your reuseable Snowpark environment.

    ```text
    psycopg2-binary
    snowflake_snowpark_python[pandas]==1.5.1
    virtualenv
    /tmp/astro_provider_snowflake-0.0.0-py3-none-any.whl
    ```

> 3. Change the content of the `Dockerfile` of your Astro project to the following, importing the `whl` file and creating a virtual environment using the [Astro venv buildkit](https://github.com/astronomer/astro-provider-venv), in which the requirements added in the previous step will be installed.

Reviewer: I think the numbering is off here - there are two step 3's

:::

6. To create an [Airflow connection](connections.md) to Snowflake and [allow serialization of Astro Python SDK objects](https://astro-sdk-python.readthedocs.io/en/stable/guides/xcom_backend.html#airflow-s-xcom-backend), add the following to your `.env` file. Make sure to enter your own Snowflake credentials as well as the name of an existing database and schema.

    ```text
    AIRFLOW__CORE__ALLOWED_DESERIALIZATION_CLASSES=airflow\.* astro\.*
    AIRFLOW__CORE__XCOM_SNOWFLAKE_CONN_NAME='snowflake_default'
    AIRFLOW_CONN_SNOWFLAKE_DEFAULT='{
        "conn_type": "snowflake",
        "login": "<username>",
        "password": "<password>",
        "schema": "MY_SKI_DATA_SCHEMA",
        "extra":
            {
                "account": "<account>",
                "warehouse": "<warehouse>",
                "database": "MY_SKI_DATA_DATABASE",
                "region": "<region>",
                "role": "<role>",
                "authenticator": "snowflake",
                "session_parameters": null,
                "application": "AIRFLOW"
            }
        }'
    ```

> 7. Optional. If you want to use a Snowflake custom XCom backend add the following variables to your `.env` as well. Make sure to replace the values with the name of your own database, schema, table and stage if you are not using the suggested values.

Reviewer: ```suggestion
7. Optional. If you want to use a Snowflake custom XCom backend, add the following additional variables to your `.env`. Replace the values with the name of your own database, schema, table, and stage if you are not using the suggested values.
```

    ```text
    AIRFLOW__CORE__XCOM_BACKEND='astronomer.providers.snowflake.xcom_backends.snowflake.SnowflakeXComBackend'
    AIRFLOW__CORE__XCOM_SNOWFLAKE_TABLE='SNOWPARK_XCOM_DB.SNOWPARK_XCOM_SCHEMA.XCOM_TABLE'
    AIRFLOW__CORE__XCOM_SNOWFLAKE_STAGE='SNOWPARK_XCOM_DB.SNOWPARK_XCOM_SCHEMA.XCOM_STAGE'
    ```

## Step 2: Add your data

> The DAG in this tutorial runs a classification model on synthetic data to predict with afternoon beverage a skier will choose based on attributes like ski color, ski resort and amount of new snow. The data is being generated using [this script](https://github.com/astronomer/airflow-snowpark-tutorial/blob/main/include/data/create_ski_dataset.py). 

Reviewer: ```suggestion
The DAG in this tutorial runs a classification model on synthetic data to predict which afternoon beverage a skier will choose based on attributes like ski color, ski resort, and amount of new snow. The data is generated using [this script](https://github.com/astronomer/airflow-snowpark-tutorial/blob/main/include/data/create_ski_dataset.py). 
```

1. Create a new directory in your Astro project's `include` directory called `data`.

2. Download the dataset from [Astronomer's GitHub](https://github.com/astronomer/learn-tutorials-data/blob/main/ski_dataset.csv) and save it in `include/data`.

## Step 3: Create your DAG

1. In your `dags` folder, create a file called `airflow_with_snowpark_tutorial.py`.

2. Copy the following code into the file. Make sure to provide your Snowflake database and schema names to `MY_SNOWFLAKE_DATABASE` and `MY_SNOWFLAKE_SCHEMA`. 

>     If you do not want the DAG to automatically setup and clean up the Snowflake custom XCom backend, set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `False` to remove the `create_snowflake_objects` and `cleanup_xcom_table` tasks from your DAG. 

Reviewer: Would it make sense to default this to False so that the simplest option is the default?

>     If you do not want the DAG to automatically setup and clean up the Snowflake custom XCom backend, set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `False` to remove the `create_snowflake_objects` and `cleanup_xcom_table` tasks from your DAG. 

Reviewer: Let's put all the DAG modification instructions in bullets in a sub-step 3.

    - `cleanup_xcom_table`: Cleans up the Snowflake custom XCom backend by dropping the `XCOM_TABLE` and `XCOM_STAGE`. This task is a [teardown task](airflow-setup-teardown.md).

## Step 4: Run your DAG

1. Run `astro dev start` in your Astro project to start up Airflow and open the Airflow UI at `localhost:8080`.

2. In the Airflow UI, run the `airflow_with_snowpark_tutorial` DAG by clicking the play button.

    ![Screenshot of the Airflow UI showing the `airflow_with_snowpark_tutorial` DAG having completed successfully in the Grid view with the Graph tab selected.](/img/tutorials/airflow-snowpark_dag_graph.png)

3. In the Snowflake UI, view the model registry to see the model that was created by the DAG. In a production context, you could pull a specific model from the registry to run predictions on new data.

    ![Screenshot of the Snowflake UI showing the model registry containing one model.](/img/tutorials/airflow-snowpark_model_registry.png)

4. Navigate to your `include` directory to view the `metrics.png` image that contains the model performance metrics shown at the start of this tutorial.

## Conclusion

> Congratulations! You trained a classification model in Snowpark using Airflow. This pipeline is meant as a simple example showcasing the three main options to run code in different Snowpark using Airflow decorators:

Reviewer: ```suggestion
Congratulations! You trained a classification model in Snowpark using Airflow. This pipeline is meant as a simple example showcasing the three main options to run code in Snowpark using Airflow decorators:
```

- `@task.snowpark_python` runs your code in a standard Snowpark environment. Use this decorator if you need to run code in Snowpark that does not require any additional packages, which are not preinstalled in a standard Snowpark environment. The corresponding traditional operator is the SnowparkPythonOperator.
- `@task.snowpark_ext_python` runs your code in a pre-existing virtual environment within Snowpark. Use this decorator when you want to reuse virtual environments in different tasks in the same Airflow instances, or your virtual environment takes a long time to be built. The corresponding traditional operator is the SnowparkExternalPythonOperator.

> - `@task.snowpark_virtualenv` runs your code in a virtual environment within Snowpark that is created on the fly just for a specific task. Use this decorator when you want to tailor a virtual environment to a task and don't need to reuse it. The corresponding traditional operator is the SnowparkVirtualenvOperator.

Reviewer: ```suggestion
- `@task.snowpark_virtualenv` runs your code in a virtual environment within Snowpark that is created at runtime for that specific task. Use this decorator when you want to tailor a virtual environment to a task and don't need to reuse it. The corresponding traditional operator is the SnowparkVirtualenvOperator.
```

---
title: "Orchestrate Snowpark Machine Learning Workflows with Apache Airflow"
sidebar_label: "Snowpark Tutorial"
description: "Learn how to integrate Snowpark and Airflow."
id: airflow-snowpark
sidebar_custom_props: { icon: 'img/integrations/snowflake.png' }
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';
import airflow_with_snowpark_tutorial from '!!raw-loader!../code-samples/dags/airflow-snowpark/airflow_with_snowpark_tutorial.py';

[Snowpark](https://www.snowflake.com/en/data-cloud/snowpark/) is a framework containing runtimes and libraries to run non-SQL code in [Snowflake](https://www.snowflake.com/en/). Snowpark comes with a [comprehensive machine learning library](https://docs.snowflake.com/en/developer-guide/snowpark-ml/index) optimized for Snowflake. 

In this tutorial, you'll learn how to: 

- Create a [custom XCom backend](custom-xcom-backends-tutorial.md) in Snowflake.
- Create and use a model registry in Snowflake.
- Use Airflow decorators to run code in Snowpark, both in a pre-built and custom virtual environment.

> - Run a [Logistic Regression model](https://mlu-explain.github.io/logistic-regression/) on a synthetic dataset describing skiers to predict their afternoon beverage choice.

Reviewer: ```suggestion
- Run a [Logistic Regression model](https://mlu-explain.github.io/logistic-regression/) on a synthetic dataset to predict skiers' afternoon beverage choice.
```

:::caution

The provider used in this tutorial is currently in beta and its contents as well as the decorators used are subject to change. After the official release, this tutorial will be updated.

:::

![A plot showing a confusion matrix and ROC curve with a high AUC for hot_chocolate, a medium AUC for snow_mocha and tea and low predictive power for wine and coffee.](/img/tutorials/airflow-snowpark_plot_metrics.png)

## Why use Airflow with Snowpark?

Snowpark allows you to use Python to perform transformations and machine learning operations on data stored in Snowflake.

Integrating Snowpark with Airflow offers the benefits of:

> - Running machine learning models directly in Snowflake, without having to move data out of your Snowflake database.

Reviewer 2: Add some bullet points here:
-Using Airflow for Snowpark python orchestration enables not only automation but also auditing, logging, retry and complex triggering for powerful workflows.
-The Snowpark provider for Airflow simplifies:
 -- connections to Snowflake without the need to pass credentials
 -- Snowpark session instantiation without repeated boilerplate code 
 -- Passing Snowpark Dataframes between tasks with automatic serialization and deserialization
 -- Data lineage 
 -- Secure data passing with a custom XCOM backend

Additionally, this tutorial shows how to use Snowflake as a custom XCom backend. This is especially useful for organizations with strict compliance requirements who want to keep all their data in Snowflake but still leverage [Airflow XCom](airflow-passing-data-between-tasks.md) to pass data between tasks.

## Time to complete

This tutorial takes approximately 45 minutes to complete.

## Assumed knowledge

To get the most out of this tutorial, make sure you have an understanding of:

- The basics of Snowflake and Snowpark. See [Introduction to Snowflake](https://docs.snowflake.com/en/user-guide-intro.html) and the [Snowpark API documentation](https://docs.snowflake.com/en/developer-guide/snowpark/index).
- Airflow decorators. [Introduction to the TaskFlow API and Airflow decorators](airflow-decorators.md).
- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).
- Setup/ teardown tasks in Airflow. See [Use setup and teardown tasks in Airflow](airflow-setup-teardown.md).

## Prerequisites

- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).

> - A Snowflake account. A [30-day free trial](https://signup.snowflake.com/) is available. You will need to have at least one database and one schema created to store the data and models used in this tutorial.

Reviewer 2: https://trial.snowflake.com/?owner=SPN-PID-365384

> - Optional. This tutorial includes instructions on how to use the Snowflake [custom XCom backend](custom-xcom-backends-tutorial.md) included in the provider. If you want to this custom XCom backend you will need to either:

Reviewer 2: If you want to use this custom XCom backend you will need to either:
    - Run the DAG with a Snowflake account with permissions to create databases, schemas, stages and tables. 
 In this case the Snowflake custom XCom backend can be created and cleaned up by tasks in the DAG itself. See [Step 3.3](#step-3-create-your-dag) for more instructions. The tutorial specifies the `ACCOUNTADMIN` privilege as it is setup automatically with the Snowflake trial.
    - Ask your Snowflake administrator to create the following for you:
        - A database called `SNOWPARK_XCOM_DB`.
        - A schema called `SNOWPARK_XCOM_SCHEMA`.
        - A table called `XCOM_TABLE`.
        - A stage called `XCOM_STAGE`.

    - Use a Snowflake account with `ACCOUNTADMIN` privileges. In this case the Snowflake custom XCom backend can be created and cleaned up by tasks in the DAG itself. See [Step 3.3](#step-3-create-your-dag) for more instructions. The free trial account does have the required privileges.

>     - Ask your Snowflake administrator to create the following for you:

Reviewer 2: Should update this whole section.  The tutorial was setup with a Trial account in mind where no objects exist already.  Alternatively they can use an existing db, schema and stage.  The only thing they need to create (or have an admin create if they don't have permissions) is the XCOM table.  If they, or an admin, create the table manually they will need the schema as specified at https://github.com/astronomer/airflow-snowpark-tutorial/blob/bfe8a8449c44c9735b16a540a02376b15e08d76e/dags/airflow_with_snowpark_tutorial.py#L101

Reviewer 2: Also need to mention that if they change the object names here they need to note them for step 8

        - A database called `SNOWPARK_XCOM_DB`.
        - A schema called `SNOWPARK_XCOM_SCHEMA`.
        - A table called `XCOM_TABLE`.
        - A stage called `XCOM_STAGE`.

:::info

The example code from this tutorial is also available on [GitHub](https://github.com/astronomer/airflow-snowpark-tutorial). 

:::

## Step 1: Configure your Astro project

1. Create a new Astro project:

    ```sh
    $ mkdir astro-snowpark-tutorial && cd astro-snowpark-tutorial
    $ astro dev init
    ```

2. Download the `whl` file for the Astro Snowflake provider beta version from the [Astronomer Github repository](https://github.com/astronomer/learn-tutorials-data/blob/main/wheel_files/astro_provider_snowflake-0.0.0-py3-none-any.whl) and save it in your Astro project's `include` directory.

3. Create a new file in your Astro project's root directory called `requirements-snowpark.txt`. This file contains all Python packages that you will install in your reuseable Snowpark environment.

    ```text
    psycopg2-binary
    snowflake_snowpark_python[pandas]==1.5.1
    virtualenv
    /tmp/astro_provider_snowflake-0.0.0-py3-none-any.whl
    ```

> 4. Change the content of the `Dockerfile` of your Astro project to the following, importing the `whl` file and creating a virtual environment using the [Astro venv buildkit](https://github.com/astronomer/astro-provider-venv), in which the requirements added in the previous step will be installed.

Reviewer 2: Add a note:

This tutorial shows how to run Snowpark Python tasks in virtual environments to simplify dependency management if needed in a production system.  In this step we create the virtual environments with ...

3. Optional. This DAG has two optional features you can enable.

    - If you want to use [setup/ teardown tasks](airflow-setup-teardown.md) to create and clean up a Snowflake custom XCom backend for this DAG, set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. This will add the `create_snowflake_objects` and `cleanup_xcom_table` tasks to your DAG and create a setup/ teardown workflow. Note that your Snowflake account needs to have `ACCOUNTADMIN` privileges to perform the operations in the `create_snowflake_objects` task and you will need to define the environment variables described in [Step 1.8](#step-1-configure-your-astro-project) to enable the custom XCom backend.

    - If you want to use a [Snowpark-optimized warehouse](https://docs.snowflake.com/en/user-guide/warehouses-snowpark-optimized) for model training, set the `USE_SNOWPARK_WH` variable to `True` and provide your warehouse names to `MY_SNOWPARK_WAREHOUSE` and `MY_SNOWFLAKE_REGULAR_WAREHOUSE`. The `MY_SNOWPARK_WAREHOUSE` warehouse will be created by the `create_snowflake_objects` task if it is enabled, otherwise you will need to create the warehouse manually before running the DAG.

:::info

> While this tutorial DAG uses a small dataset where model training can be accomplished using the standard Snowflake warehouse, Astronomer recommends to use a Snowpark-optimized warehouse for model training in production.

Reviewer: ```suggestion
While this tutorial DAG uses a small dataset where model training can be accomplished using the standard Snowflake warehouse, Astronomer recommends using a Snowpark-optimized warehouse for model training in production.
```

---
title: "Orchestrate Snowpark Machine Learning Workflows with Apache Airflow"
sidebar_label: "Snowpark Tutorial"
description: "Learn how to integrate Snowpark and Airflow."
id: airflow-snowpark
sidebar_custom_props: { icon: 'img/integrations/snowflake.png' }
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';
import airflow_with_snowpark_tutorial from '!!raw-loader!../code-samples/dags/airflow-snowpark/airflow_with_snowpark_tutorial.py';

> [Snowpark](https://www.snowflake.com/en/data-cloud/snowpark/) is a framework containing runtimes and libraries to run non-SQL code in [Snowflake](https://www.snowflake.com/en/). Snowpark comes with a [comprehensive machine learning library](https://docs.snowflake.com/en/developer-guide/snowpark-ml/index) optimized for Snowflake. 

Reviewer 3: ```suggestion
[Snowpark](https://www.snowflake.com/en/data-cloud/snowpark/) is a framework that contains runtimes and libraries to run non-SQL code in [Snowflake](https://www.snowflake.com/en/). Snowpark comes with a [comprehensive machine learning library](https://docs.snowflake.com/en/developer-guide/snowpark-ml/index) optimized for Snowflake. 
```

In this tutorial, you'll learn how to: 

- Create a [custom XCom backend](custom-xcom-backends-tutorial.md) in Snowflake.
- Create and use a model registry in Snowflake.
- Use Airflow decorators to run code in Snowpark, both in a pre-built and custom virtual environment.

> - Run a [Logistic Regression model](https://mlu-explain.github.io/logistic-regression/) on a synthetic dataset to predict skiers' afternoon beverage choice.

Reviewer 3: This is fabulous.

:::caution

> The provider used in this tutorial is currently in beta and its contents as well as the decorators used are subject to change. After the official release, this tutorial will be updated.

Reviewer 3: ```suggestion
The provider used in this tutorial is currently in beta and both its contents and decorators are subject to change. After the official release, this tutorial will be updated.
```

:::

> ![A plot showing a confusion matrix and ROC curve with a high AUC for hot_chocolate, a medium AUC for snow_mocha and tea and low predictive power for wine and coffee.](/img/tutorials/airflow-snowpark_plot_metrics.png)

Reviewer 3: ```suggestion
![A plot showing a confusion matrix and ROC curve with a high AUC for hot_chocolate, a medium AUC for snow_mocha and tea, and low predictive power for wine and coffee.](/img/tutorials/airflow-snowpark_plot_metrics.png)
```

## Why use Airflow with Snowpark?

Snowpark allows you to use Python to perform transformations and machine learning operations on data stored in Snowflake.

Integrating Snowpark with Airflow offers the benefits of:

- Running machine learning models directly in Snowflake, without having to move data out of your Snowflake database.

> - Express data transformations in Snowflake in Python instead of SQL.
> - Store and version your machine learning models in a model registry inside Snowflake.
> - Use Snowpark's compute resources instead of the resources of your Airflow cluster for machine learning.
> - Use Airflow for Snowpark Python orchestration to enable automation, auditing, logging, and retry and complex triggering for powerful workflows.

Reviewer 3: ```suggestion
- Expressing data transformations in Snowflake in Python instead of SQL.
- Storing and versioning your machine learning models in a model registry inside Snowflake.
- Using Snowpark's compute resources instead of your Airflow cluster resources for machine learning.
- Using Airflow for Snowpark Python orchestration to enable automation, auditing, logging, retry, and complex triggering for powerful workflows.
```

> - Using Airflow for Snowpark python orchestration enables not only automation but also auditing, logging, retry and complex triggering for powerful workflows.

Reviewer: ```suggestion
- Use Airflow for Snowpark Python orchestration to enable automation, auditing, logging, and retry and complex triggering for powerful workflows.
```

The Snowpark provider for Airflow simplifies interacting with Snowpark by:

- Connecting to Snowflake using an [Airflow connection](connections/snowflake.md), removing the need to directly pass credentials in your DAG.
- Automatically instantiating a Snowpark session.
- Automatically serializing and deserializing Snowpark dataframes passed using [Airflow XCom](airflow-passing-data-between-tasks.md).
- Integrating with [OpenLineage](airflow-openlineage.md).
- Providing a pre-built custom XCom backend for Snowflake.

> Additionally, this tutorial shows how to use Snowflake as a custom XCom backend. This is especially useful for organizations with strict compliance requirements who want to keep all their data in Snowflake but still leverage [Airflow XCom](airflow-passing-data-between-tasks.md) to pass data between tasks.

Reviewer 3: ```suggestion
Additionally, this tutorial shows how to use Snowflake as a custom XCom backend. This is especially useful for organizations with strict compliance requirements who want to keep all their data in Snowflake, but still leverage [Airflow XCom](airflow-passing-data-between-tasks.md) to pass data between tasks.
```

## Time to complete

This tutorial takes approximately 45 minutes to complete.

## Assumed knowledge

To get the most out of this tutorial, make sure you have an understanding of:

- The basics of Snowflake and Snowpark. See [Introduction to Snowflake](https://docs.snowflake.com/en/user-guide-intro.html) and the [Snowpark API documentation](https://docs.snowflake.com/en/developer-guide/snowpark/index).

> - Airflow decorators. [Introduction to the TaskFlow API and Airflow decorators](airflow-decorators.md).

Reviewer 3: ```suggestion
- Airflow decorators. See [Introduction to the TaskFlow API and Airflow decorators](airflow-decorators.md).
```

- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).
- Setup/ teardown tasks in Airflow. See [Use setup and teardown tasks in Airflow](airflow-setup-teardown.md).

## Prerequisites

- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).

> - A Snowflake account. A [30-day free trial](https://trial.snowflake.com/?owner=SPN-PID-365384) is available. You will need to have at least one database and one schema created to store the data and models used in this tutorial.

Reviewer 3: ```suggestion
- A Snowflake account. A [30-day free trial](https://trial.snowflake.com/?owner=SPN-PID-365384) is available. You need to have at least one database and one schema created to store the data and models used in this tutorial.
```

> - Optional. This tutorial includes instructions on how to use the Snowflake [custom XCom backend](custom-xcom-backends-tutorial.md) included in the provider. If you want to this custom XCom backend you will need to either:

Reviewer 3: ```suggestion
- (Optional) This tutorial includes instructions on how to use the Snowflake [custom XCom backend](custom-xcom-backends-tutorial.md) included in the provider. If you want to this custom XCom backend you will need to either:
```

We changed this styling a little while back for the docs. Open to discussion about having same/different conventions between the two!

Author: Thanks for the headsup! I vote with using the same conventions whereever possible between Docs and Learn :) 
Will keep it in mind.

>     - Run the DAG using a Snowflake account with `ACCOUNTADMIN` privileges to let the first task of the DAG create the required database, schema, stage and table. See [Step 3.3](#step-3-create-your-dag) for more instructions. The free trial account does have the required privileges.

Reviewer 3: ```suggestion
    - Run the DAG using a Snowflake account with `ACCOUNTADMIN` privileges to allow the DAG's first task create the required database, schema, stage and table. See [Step 3.3](#step-3-create-your-dag) for more instructions. The free trial account has the required privileges.
```

    - Ask your Snowflake administrator to: 

>         - Provider you with the name of an existing database, schema and stage. You will need to use these names in [Step 1.8](#step-1-configure-your-astro-project) for the `AIRFLOW__CORE__XCOM_SNOWFLAKE_TABLE` and `AIRFLOW__CORE__XCOM_SNOWFLAKE_STAGE` environment variables. 

Reviewer 3: ```suggestion
        - Provide you with the name of an existing database, schema, and stage. You need to use these names in [Step 1.8](#step-1-configure-your-astro-project) for the `AIRFLOW__CORE__XCOM_SNOWFLAKE_TABLE` and `AIRFLOW__CORE__XCOM_SNOWFLAKE_STAGE` environment variables. 
```

        - Create an `XCOM_TABLE` with the following schema:

        ```sql
        dag_id varchar NOT NULL, 
        task_id varchar NOT NULL, 
        run_id varchar NOT NULL,
        multi_index integer NOT NULL,
        key varchar NOT NULL,
        value_type varchar NOT NULL,
        value varchar NOT NULL
        ```

:::info

The example code from this tutorial is also available on [GitHub](https://github.com/astronomer/airflow-snowpark-tutorial). 

:::

## Step 1: Configure your Astro project

1. Create a new Astro project:

    ```sh
    $ mkdir astro-snowpark-tutorial && cd astro-snowpark-tutorial
    $ astro dev init
    ```

2. Download the `whl` file for the Astro Snowflake provider beta version from the [Astronomer Github repository](https://github.com/astronomer/learn-tutorials-data/blob/main/wheel_files/astro_provider_snowflake-0.0.0-py3-none-any.whl) and save it in your Astro project's `include` directory.

> 3. Create a new file in your Astro project's root directory called `requirements-snowpark.txt`. This file contains all Python packages that you will install in your reuseable Snowpark environment.

Reviewer 3: ```suggestion
3. Create a new file in your Astro project's root directory called `requirements-snowpark.txt`. This file contains all Python packages that you install in your reuseable Snowpark environment.
```

    ```text
    psycopg2-binary
    snowflake_snowpark_python[pandas]==1.5.1
    virtualenv
    /tmp/astro_provider_snowflake-0.0.0-py3-none-any.whl
    ```

> 4. Change the content of the `Dockerfile` of your Astro project to the following, importing the `whl` file and creating a virtual environment using the [Astro venv buildkit](https://github.com/astronomer/astro-provider-venv), in which the requirements added in the previous step will be installed. This tutorial includes Snowpark Python tasks that are running in virtual environment, which is a common pattern in production to simplify dependency management. This Dockerfile creates a virtual environment called `snowpark` with the Python version 3.8 and the packages specified in `requirements-snowpark.txt`. 

Reviewer 3: ```suggestion
4. Change the content of the `Dockerfile` of your Astro project to the following, which imports the `whl` file and creates a virtual environment by using the [Astro venv buildkit](https://github.com/astronomer/astro-provider-venv). The requirements added in the previous step are installed in that virtual environment. This tutorial includes Snowpark Python tasks that are running in virtual environment, which is a common pattern in production to simplify dependency management. This Dockerfile creates a virtual environment called `snowpark` with the Python version 3.8 and the packages specified in `requirements-snowpark.txt`. 
```
I might need to take another pass at editing this step.

:::

7. To create an [Airflow connection](connections.md) to Snowflake and [allow serialization of Astro Python SDK objects](https://astro-sdk-python.readthedocs.io/en/stable/guides/xcom_backend.html#airflow-s-xcom-backend), add the following to your `.env` file. Make sure to enter your own Snowflake credentials as well as the name of an existing database and schema.

    ```text
    AIRFLOW__CORE__ALLOWED_DESERIALIZATION_CLASSES=airflow\.* astro\.*
    AIRFLOW__CORE__XCOM_SNOWFLAKE_CONN_NAME='snowflake_default'
    AIRFLOW_CONN_SNOWFLAKE_DEFAULT='{
        "conn_type": "snowflake",
        "login": "<username>",
        "password": "<password>",
        "schema": "MY_SKI_DATA_SCHEMA",
        "extra":
            {
                "account": "<account>",
                "warehouse": "<warehouse>",
                "database": "MY_SKI_DATA_DATABASE",
                "region": "<region>",
                "role": "<role>",
                "authenticator": "snowflake",
                "session_parameters": null,
                "application": "AIRFLOW"
            }
        }'
    ```

> 8. Optional. If you want to use a Snowflake custom XCom backend, add the following additional variables to your `.env`. Replace the values with the name of your own database, schema, table, and stage if you are not using the suggested values.

Reviewer 3: ```suggestion
8. (Optional) If you want to use a Snowflake custom XCom backend, add the following additional variables to your `.env`. Replace the values with the name of your own database, schema, table, and stage if you are not using the suggested values.
```

    AIRFLOW__CORE__XCOM_SNOWFLAKE_STAGE='SNOWPARK_XCOM_DB.SNOWPARK_XCOM_SCHEMA.XCOM_STAGE'
    ```

## Step 2: Add your data

The DAG in this tutorial runs a classification model on synthetic data to predict which afternoon beverage a skier will choose based on attributes like ski color, ski resort, and amount of new snow. The data is generated using [this script](https://github.com/astronomer/airflow-snowpark-tutorial/blob/main/include/data/create_ski_dataset.py). 

1. Create a new directory in your Astro project's `include` directory called `data`.

2. Download the dataset from [Astronomer's GitHub](https://github.com/astronomer/learn-tutorials-data/blob/main/ski_dataset.csv) and save it in `include/data`.

## Step 3: Create your DAG

1. In your `dags` folder, create a file called `airflow_with_snowpark_tutorial.py`.

2. Copy the following code into the file. Make sure to provide your Snowflake database and schema names to `MY_SNOWFLAKE_DATABASE` and `MY_SNOWFLAKE_SCHEMA`. 

    <CodeBlock language="python">{airflow_with_snowpark_tutorial}</CodeBlock>

    This DAG consists of eight tasks in a simple ML orchestration pipeline.

>     - Optional. `create_snowflake_objects`: Creates the Snowflake objects required for the Snowflake custom XCom backend. This task uses the `@task.snowflake_python` decorator to run code within Snowpark, automatically instantiating a Snowpark session called `snowpark_session` from the connection ID provided to the `snowflake_conn_id` parameter. This task is a [setup task](airflow-setup-teardown.md) and will only be shown in the DAG graph if you set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. See also Step 3.3.

Reviewer 3: ```suggestion
    - (Optional) `create_snowflake_objects`: Creates the Snowflake objects required for the Snowflake custom XCom backend. This task uses the `@task.snowflake_python` decorator to run code within Snowpark, automatically instantiating a Snowpark session called `snowpark_session` from the connection ID provided to the `snowflake_conn_id` parameter. This task is a [setup task](airflow-setup-teardown.md) and is only shown in the DAG graph if you set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. See also Step 3.3.
```

    - `load_file`: Loads the data from the `ski_dataset.csv` file into the Snowflake table `MY_SNOWFLAKE_TABLE` using the [load_file operator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html) from the Astro Python SDK. 

>     - `create_model_registry`: Creates a model registry in Snowpark using the [Snowpark ML package](https://docs.snowflake.com/en/developer-guide/snowpark-ml/index). Since the task is defined using the `@task.snowflake_python` decorator, the snowpark session is automatically instantiated from provided connection ID.

Reviewer 3: ```suggestion
    - `create_model_registry`: Creates a model registry in Snowpark using the [Snowpark ML package](https://docs.snowflake.com/en/developer-guide/snowpark-ml/index). Since the task is defined by the `@task.snowflake_python` decorator, the snowpark session is automatically instantiated from provided connection ID.
```

    - `transform_table_step_one`: Transforms the data in the Snowflake table using Snowpark syntax to filter to only include rows of skiers that ordered the beverages we are interested in. Computation of this task runs within Snowpark. The resulting table is written to [XCom](airflow-passing-data-between-tasks.md) as a pandas DataFrame. 

    - `transform_table_step_two`: Transforms the pandas dataframe created by the upstream task to filter only for serious skiers (those who skied at least one hour that day).

>     This task uses the `@task.snowpark_ext_python` decorator, running the code in the Snowpark virtual environment created in Step 1. Which virtual environment to run a task in is determined by the binary provided to the `python` parameter of the decorator. The `@task.snowpark_ext_python` decorator works analogously to the [@task.external_python decorator](external-python-operator.md) with the difference being that the code is executed within Snowpark's compute.

Reviewer 3: ```suggestion
    This task uses the `@task.snowpark_ext_python` decorator, running the code in the Snowpark virtual environment created in Step 1. The binary provided to the `python` parameter of the decorator determines which virtual environment to run a task in. The `@task.snowpark_ext_python` decorator works analogously to the [@task.external_python decorator](external-python-operator.md), except the code is executed within Snowpark's compute.
```

>     - `train_beverage_classifier`: Trains a [Snowpark Logistic Regression model](https://docs.snowflake.com/en/developer-guide/snowpark-ml/reference/latest/api/modeling/snowflake.ml.modeling.linear_model.LogisticRegression) on the dataset, saves the model to the model registry and creates predictions from a test dataset. This task uses the `@task.snowpark_virtualenv` decorator to run the code in a newly created virtual environment within Snowpark's compute. The packages to be installed within the virtual environment are specified to the `requirements` parameter of the decorator. The model predictions are saved to XCom as a pandas DataFrame.

Reviewer 3: ```suggestion
    - `train_beverage_classifier`: Trains a [Snowpark Logistic Regression model](https://docs.snowflake.com/en/developer-guide/snowpark-ml/reference/latest/api/modeling/snowflake.ml.modeling.linear_model.LogisticRegression) on the dataset, saves the model to the model registry, and creates predictions from a test dataset. This task uses the `@task.snowpark_virtualenv` decorator to run the code in a newly created virtual environment within Snowpark's compute. The `requirements` parameter of the decorator specifies the packages to install in the virtual environment. The model predictions are saved to XCom as a pandas DataFrame.
```

    - `plot_metrics`: Creates a plot of the model performance metrics and saves it to the `include` directory. This task runs in the Airflow environment using the `@task` decorator.

>     - Optional. `cleanup_xcom_table`: Cleans up the Snowflake custom XCom backend by dropping the `XCOM_TABLE` and `XCOM_STAGE`. This task is a [teardown task](airflow-setup-teardown.md) and will only be shown in the DAG graph if you set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. See also Step 3.3.

Reviewer 3: ```suggestion
    - (Optional) `cleanup_xcom_table`: Cleans up the Snowflake custom XCom backend by dropping the `XCOM_TABLE` and `XCOM_STAGE`. This task is a [teardown task](airflow-setup-teardown.md) and is only shown in the DAG graph if you set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. See also Step 3.3.
```

3. Optional. This DAG has two optional features you can enable.

>     - If you want to use [setup/ teardown tasks](airflow-setup-teardown.md) to create and clean up a Snowflake custom XCom backend for this DAG, set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. This will add the `create_snowflake_objects` and `cleanup_xcom_table` tasks to your DAG and create a setup/ teardown workflow. Note that your Snowflake account needs to have `ACCOUNTADMIN` privileges to perform the operations in the `create_snowflake_objects` task and you will need to define the environment variables described in [Step 1.8](#step-1-configure-your-astro-project) to enable the custom XCom backend.

Reviewer 3: ```suggestion
    - If you want to use [setup/ teardown tasks](airflow-setup-teardown.md) to create and clean up a Snowflake custom XCom backend for this DAG, set `SETUP_TEARDOWN_SNOWFLAKE_CUSTOM_XCOM_BACKEND` to `True`. This setting adds the `create_snowflake_objects` and `cleanup_xcom_table` tasks to your DAG and creates a setup/ teardown workflow. Note that your Snowflake account needs to have `ACCOUNTADMIN` privileges to perform the operations in the `create_snowflake_objects` task and you need to define the environment variables described in [Step 1.8](#step-1-configure-your-astro-project) to enable the custom XCom backend.
```

this is a quirk of a lot of technical writing. As someone reads through a procedure, every step is a snapshot of their present moment. In step 3 they are completing that task in the present moment. When there are conditional situations, it's usually clearest to structure that as "If [you do x], then y occurs.", because other formats of conditional future can be a really tricky verb construction that slows people down to think through. 

"If [you do x], then y occurs." vs 'When you do x, y will occur"-> when in the future will y occur? At this step? At the end of the process?

Author: Thank you for the explanation, that does make a lot of sense! (and I've definitely read docs before, performed a step and was left wondering if y was supposed to occur later or I did something wrong 😅 )

>     - If you want to use a [Snowpark-optimized warehouse](https://docs.snowflake.com/en/user-guide/warehouses-snowpark-optimized) for model training, set the `USE_SNOWPARK_WH` variable to `True` and provide your warehouse names to `MY_SNOWPARK_WAREHOUSE` and `MY_SNOWFLAKE_REGULAR_WAREHOUSE`. The `MY_SNOWPARK_WAREHOUSE` warehouse will be created by the `create_snowflake_objects` task if it is enabled, otherwise you will need to create the warehouse manually before running the DAG.

Reviewer 3: ```suggestion
    - If you want to use a [Snowpark-optimized warehouse](https://docs.snowflake.com/en/user-guide/warehouses-snowpark-optimized) for model training, set the `USE_SNOWPARK_WH` variable to `True` and provide your warehouse names to `MY_SNOWPARK_WAREHOUSE` and `MY_SNOWFLAKE_REGULAR_WAREHOUSE`. If the `create_snowflake_objects` task is enabled, it creates the `MY_SNOWPARK_WAREHOUSE` warehouse. Otherwise, you need to create the warehouse manually before running the DAG.
```
